Add docs page for breaking notes

heliocliu opened the pull request:
This is to prepare for the more detailed changelogs that we expect to be publishing in the future after public preview.

Some notes:
- I added a section in the docs site as a placeholder for changelogs.  The expectation is that the section's index page will link to all the change notes pages, where each release gets its own page (image of how it looks attached)
- The existing BREAKING.md will be reviewed at release for formatting etc. and a release notes page for that release will be created under `docs/changelogs/releases/<some release identifier>`, and a link to it will be added under the main changelogs index.  This is currently manual, and future work is expected to automate adding additional information, such as released version numbers or additional individual change information
- This change doesn't address any of the automation stuff yet (soon™)

<img width="876" alt="Screen Shot 2021-09-16 at 9 07 01 AM" src="https://user-images.githubusercontent.com/59622401/133647034-21655346-5dcd-4d85-8c87-8e83274cfb4f.png">

heliocliu committed:
initial commit

heliocliu committed:
fix misplaced breaking message

heliocliu commented on docs/content/docs/changelogs/CHANGELOG.md:
```diff
@@ -0,0 +1,17 @@
+---
+title: Working changelog
+draft: true
+description: Explanations of breaking and notable changes go here.  Images should live in docs/static/images.  The contents of this file are moved to a release notes page as parts of building docs for each release.
```
unsure if we want to reuse this directory for all image resources or keep them inside this content bundle like under `docs/changelogs/images`

tylerbutler replied:
Do we need images? I'd prefer to keep binaries out of the repo if possible. For the rest of the site, we have a CDN that we can upload to, but maybe that's too much overhead for this? What do you think?

heliocliu replied:
I'm not a fan of the cdn upload unless it's going to be incredibly easy to do.  If a dev is trying to add something here with an image then suddenly needs to go through some authentication process or ask someone else to do it, I think there's a good chance the image just gets skipped.

That said I'm not sure if we'll need images either, since the product isn't really some app with UI.  I'm fine advising against binary images like you say.  I'd also be fine with a temporary solution (like putting them in the repo) and revisiting later to see how big of a problem it is

tylerbutler commented on docs/content/docs/changelogs/CHANGELOG.md:
```diff
@@ -0,0 +1,17 @@
+description: Explanations of breaking and notable changes go here.  Images should live in docs/static/images.  The contents of this file are moved to a release notes page as parts of building docs for each release.
+---
+
+## 0.48 Breaking and notable changes
+This release is the Public Preview!
```
```suggestion
## 0.48 Breaking and notable changes

This release is the Public Preview!
```

Nit: markdown is line-break sensitive, so technically without the extra line break it's all just one long header line. I think most markdown parsers handle it correctly, but I am a pedant. :)

tylerbutler replied:
Not sure the "public preview" designation is needed. Something might change, and I don't think we're using that terminology anywhere else.

heliocliu replied:
will remove the line.  wasn't aware of the spec on line breaks, good to know.  i've definitely just been doing the look-and-see approach with the parsers🤐

heliocliu replied:
```suggestion
## 0.48 Breaking and notable changes
```

tylerbutler commented on docs/content/docs/changelogs/CHANGELOG.md:
```diff
@@ -0,0 +1,17 @@
+---
+title: Working changelog
+draft: true
```
```suggestion
draft: true # Don't publish this!
```

You can change this - It's just worth noting that the "draft" is intentional for the benefit of future folks.

tylerbutler reviewed (commented):
I like the direction. Couple of higher level questions: do you imagine each release as an individual file/page, or are all releases on one page? Also, just to confirm, this proposal is to leave BREAKING.md in place, and continue to use it, but introduce a new more curated section of the docs that will have (possibly duplicate) info?

heliocliu committed:
Update docs/content/docs/changelogs/CHANGELOG.md

Co-authored-by: Tyler Butler <tyler@tylerbutler.com>

heliocliu commented:
> I like the direction. Couple of higher level questions: do you imagine each release as an individual file/page, or are all releases on one page? Also, just to confirm, this proposal is to leave BREAKING.md in place, and continue to use it, but introduce a new more curated section of the docs that will have (possibly duplicate) info?

I'm thinking an individual page for each release, where the index page links to them.  I think the releases will be frequent enough that when we factor in things like associated release versions for individual packages when they diverge, putting them all on the same page will be too much.  It also gives us more flexibility with automation and organization if we ever change the formatting.

Idea with the BREAKING.md is to only add new changes in the new location.  I'm keeping the old file because it's a lot more discoverable at the moment and people are used to editing it.  Then we can remove it later when we have some automated checks that can communicate where to add the notes, since I don't think it's very discoverable as is.  Did you think it would make more sense to leave BREAKING.md as is, and manually edit/curate to the new location at release?

heliocliu committed:
Merge remote-tracking branch 'upstream/main' into BreakingMove

heliocliu committed:
pr feedback

skylerjokiel commented:
Typescript has a what's new section on their site that I really like and think makes it easy to track changes across Minor/Major builds. @heliocliu / @tylerbutler do we have thoughts of something similar? 

https://www.typescriptlang.org/docs/handbook/release-notes/overview.html

skylerjokiel commented:
> > I like the direction. Couple of higher level questions: do you imagine each release as an individual file/page, or are all releases on one page? Also, just to confirm, this proposal is to leave BREAKING.md in place, and continue to use it, but introduce a new more curated section of the docs that will have (possibly duplicate) info?
> 
> I'm thinking an individual page for each release, where the index page links to them. I think the releases will be frequent enough that when we factor in things like associated release versions for individual packages when they diverge, putting them all on the same page will be too much. It also gives us more flexibility with automation and organization if we ever change the formatting.
> 
> Idea with the BREAKING.md is to only add new changes in the new location. I'm keeping the old file because it's a lot more discoverable at the moment and people are used to editing it. Then we can remove it later when we have some automated checks that can communicate where to add the notes, since I don't think it's very discoverable as is. Did you think it would make more sense to leave BREAKING.md as is, and manually edit/curate to the new location at release?

I'm concerned that a brain dump of all the breaking changes across the repo will be confusing for our 3P developers considering the 3P API doesn't expose a large percentage of what we consider "breaking changes" and would be added to the BREAKING.md currently. So I see a collection of changes that would be breaking for 1P consumers of the framework that would just be confusing if you tried to explain it to a 3P developer. Ideally as we continue to add features to the 3P API this set will converge but currently I don't think it's going to work.

I'm not quite sure how to tackle the problem without duplicating content but maybe this problem becomes easier if we split the azure-client package from the client build systems. 

Maybe a temp solution is to copy 3P API breaking changes as part of the release cycle, or depending on how we plan on exposing these breaks a step after. @tylerbutler I'm curious if we will want to curate this to make it understandable and present a unified voice.

heliocliu commented:
> Typescript has a what's new section on their site that I really like and think makes it easy to track changes across Minor/Major builds. @heliocliu / @tylerbutler do we have thoughts of something similar?
> 
> https://www.typescriptlang.org/docs/handbook/release-notes/overview.html

I hadn't seen theirs but the formatting for this change is very similar, except the list of releases is in the overview change (instead of the menu).  I'll add a screenshot of what it looks like here

> I'm concerned that a brain dump of all the breaking changes across the repo will be confusing for our 3P developers considering the 3P API doesn't expose a large percentage of what we consider "breaking changes" and would be added to the BREAKING.md currently. So I see a collection of changes that would be breaking for 1P consumers of the framework that would just be confusing if you tried to explain it to a 3P developer. Ideally as we continue to add features to the 3P API this set will converge but currently I don't think it's going to work.
> 
> I'm not quite sure how to tackle the problem without duplicating content but maybe this problem becomes easier if we split the azure-client package from the client build systems.
> 
> Maybe a temp solution is to copy 3P API breaking changes as part of the release cycle, or depending on how we plan on exposing these breaks a step after. @tylerbutler I'm curious if we will want to curate this to make it understandable and present a unified voice.

I think we're 100% on curation to ensure the points you noted.  I don't think it really matters where specifically the BREAKING.md is located, but I figured in the docs folder would be better because that's where it's going and so it's clearer for devs where to put any media resources.  But I don't think we want to separate 3P and 1P breaking notes because of independent versioning.  As in, we need to provided version information for all packages and how the uber-package version relates to its individual packages and it makes sense to have that with the release notes, so if the 1P consumers are already referencing this for version numbers it makes sense to put 1P breaking notes here as well (even if ultimately they are sectioned from 3P notes on the docs page).  Thoughts?

tylerbutler commented:
> Typescript has a what's new section on their site that I really like and think makes it easy to track changes across Minor/Major builds.

Yeah, I think that's a great model, and matches what I had in my head. I'm drafting release notes for .47 and .48 as a starting point -- aiming to have that done this week.

> I'm not quite sure how to tackle the problem without duplicating content but maybe this problem becomes easier if we split the azure-client package from the client build systems.

I support splitting the azure-client to be built separately, regardless of this. :) 

> Maybe a temp solution is to copy 3P API breaking changes as part of the release cycle, or depending on how we plan on exposing these breaks a step after.

> As in, we need to provided version information for all packages and how the uber-package version relates to its individual packages and it makes sense to have that with the release notes, so if the 1P consumers are already referencing this for version numbers it makes sense to put 1P breaking notes here as well (even if ultimately they are sectioned from 3P notes on the docs page).

I think the key is in grouping the release notes and hiding the core packages by default. I.e. imagine a page for 0.48 that has a section for fluid-framework and the service client packages, then an "other packages" section that is maybe collapsed -- or is on another page that's linked from the main release notes. In other words, the public API is front and center, but the core package notes are also available by clicking through.

A related option is to group the changes by the "area" -- the labels we use -- or by the layer the change affects.

tylerbutler commented:
> I don't think it really matters where specifically the BREAKING.md is located

One argument for keeping it in the root is that the docs might be split into a separate repo in the future. That said, I don't think we should make the decision based on a _possible_ future. BUT -- this did make me think that maybe we should think about aggregating release notes from individual packages. I.e. at some point with independent versioning we're going to have packages with their own release cycle and we'll want to roll up those releases in some (automated) way.

heliocliu commented:
> > I don't think it really matters where specifically the BREAKING.md is located
> 
> One argument for keeping it in the root is that the docs might be split into a separate repo in the future. That said, I don't think we should make the decision based on a _possible_ future. BUT -- this did make me think that maybe we should think about aggregating release notes from individual packages. I.e. at some point with independent versioning we're going to have packages with their own release cycle and we'll want to roll up those releases in some (automated) way.

I do like the idea of individual notes for each package because it makes it easier to tell which are for 1P and 3P libraries at edit/release, though I'd guess there are situations where it might get kind of confusing.  E.g. if you change a common interface that's exposed as an arg/return type in a lot of packages.

All that said, our breakings notes have done fine without any media attachments so far, so I'd also be fine leaving it as is (because there's the manual curation step anyway), and worrying about what to do with media when it actually happens

tylerbutler commented:
> I'd also be fine leaving it as is (because there's the manual curation step anyway), and worrying about what to do with media when it actually happens

Yeah -- agreed!

heliocliu committed:
move breaking back to old location

heliocliu committed:
Merge remote-tracking branch 'upstream/main' into BreakingMove

heliocliu commented:
@tylerbutler I've moved the breaking file back and added some tips on writing a good change note to ease the editing/curation process and avoid needing to re-consult with the developer

heliocliu committed:
Update BREAKING.md

heliocliu committed:
Merge remote-tracking branch 'upstream/main' into BreakingMove

ChumpChief reviewed (approved):
Change looks fine assuming the direction is agreed on.
